[READY] Hypospray MK IIs

Poojawa opened the pull request:
:cl: Poojawa
add: Hypospray mk IIs are being deployed to stations, these should provide an easier time for medical staff! They come with both inject and spray modes! Spray mode acts like a patch for applying meds
soundadd: Hyposprays are fancy, you'll know when they're being used.
server: Due to fucky-ness, you'll need to unload, then reload the hypos at least once. just how it be.
/:cl:

Something robust for medical staff to have, pending the nerfs to pills and patches.
Sprites are from Cameron. <3

![31865111-af4979be-b72e-11e7-89f0-e34f62becc25](https://user-images.githubusercontent.com/6293118/36710287-8e08c8fe-1b42-11e8-9080-935ad58fc90a.png)

Poojawa committed:
Initial commit

Poojawa committed:
most of the work down

itfinallycompiled

Poojawa committed:
tweaks

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into hypospray-meme

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into hypospray-meme

Poojawa committed:
ree

Poojawa committed:
fixes bad attempt at mode switching

Poojawa committed:
slight sprite tweak to give vials better clickability.

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+		start_vial = new loaded_vial(src)
+	update_icon()
+
+/obj/item/reagent_containers/hypospray/mkii/examine(mob/user)
```
Add a . = ..() below here

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+	to_chat(user, "[src] is set to [src.mode ? "Inject" : "Spray"] contents on application.")
+
+/obj/item/reagent_containers/hypospray/mkii/proc/unload_hypo(obj/item/reagent_containers/glass/bottle/vial/W, mob/user as mob)
+	if(src.contents == null)
```
No need for src here

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+		to_chat(user, "This doesn't have a vial!")
+		return
+	else
+		src.reagents.trans_to(W, src.volume)
```
No need for src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+		return
+	else
+		src.reagents.trans_to(W, src.volume)
+		src.volume = 0
```
No need for src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+		W.loc = get_turf(user.loc)
+		W.update_icon()
+		user.put_in_hands(W)
+		src.contents -= W
```
No need for src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+		src.contents -= W
+		to_chat(user, "<span class='notice'>You remove the vial from the [src].</span>")
+		update_icon()
+		playsound(src.loc, 'sound/weapons/empty.ogg', 50, 1)
```
You just need src here

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+		to_chat(user, "<span class='notice'>\The [src] only accepts hypospray vials!</span>")
+		return
+	else
+		if(src.contents.len == 1)
```
No need for src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+			user.visible_message("<span class='notice'>[user] begins loading a vial into \the [src].</span>","<span class='notice'>You start loading [W] into \the [src].</span>")
+			if(!do_after(user,30) || W || !(W in user))
+				return FALSE
+			src.contents += W
```
No need for src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+			if(!do_after(user,30) || W || !(W in user))
+				return FALSE
+			src.contents += W
+			src.volume = W.volume
```
No need for src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+			W.reagents.trans_to(src, W.volume)
+			user.visible_message("<span class='notice'>[user] has loaded a vial into \the [src].</span>","<span class='notice'>You have loaded [W] into \the [src].</span>")
+			update_icon()
+			playsound(src.loc, 'sound/weapons/autoguninsert.ogg', 50, 1)
```
You can replace src.loc with just src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+					return
+				if(L.reagents.total_volume >= L.reagents.maximum_volume)
+					return
+				if(src.possible_transfer_amounts >= 15)
```
No need for src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+				if(L.reagents.total_volume >= L.reagents.maximum_volume)
+					return
+				if(src.possible_transfer_amounts >= 15)
+					playsound(loc,'sound/items/hypospray_long.ogg',50, 1, -1)
```
Replace loc with src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+				if(src.possible_transfer_amounts >= 15)
+					playsound(loc,'sound/items/hypospray_long.ogg',50, 1, -1)
+				else
+					playsound(loc, pick(src.inject_sounds), 50, 1, -1)
```
Replace loc with src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+			if(L != user)
+				add_logs(user, L, "injected", src, addition="which had [contained]")
+			else
+				if(src.possible_transfer_amounts >= 15)
```
No need for src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+					playsound(loc,'sound/items/hypospray_long.ogg',50, 1, -1)
+				else
+					playsound(loc, pick(src.inject_sounds), 50, 1, -1)
+				log_attack("<font color='red'>[user.name] ([user.ckey]) injected [L.name] ([L.ckey]) with [src.name], which had [contained] (INTENT: [uppertext(user.a_intent)])</font>")
```
Replace src.name with just src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+				if(src.possible_transfer_amounts >= 15)
+					playsound(loc,'sound/items/hypospray_long.ogg',50, 1, -1)
+				else
+					playsound(loc, pick(src.inject_sounds), 50, 1, -1)
```
Replace src.inject_sounds with just inject_sounds, replace loc with src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+				add_logs(user, L, "injected", src, addition="which had [contained]")
+			else
+				if(src.possible_transfer_amounts >= 15)
+					playsound(loc,'sound/items/hypospray_long.ogg',50, 1, -1)
```
Replace loc with src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+					return
+				if(L.reagents.total_volume >= L.reagents.maximum_volume)
+					return
+				if(src.possible_transfer_amounts >= 15)
```
No need for src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+				if(L.reagents.total_volume >= L.reagents.maximum_volume)
+					return
+				if(src.possible_transfer_amounts >= 15)
+					playsound(loc,'sound/items/hypospray_long.ogg',50, 1, -1)
```
Replace loc with src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+				if(src.possible_transfer_amounts >= 15)
+					playsound(loc,'sound/items/hypospray_long.ogg',50, 1, -1)
+				else
+					playsound(loc, pick(src.inject_sounds), 50, 1, -1)
```
Replace loc with src, you can replace src.inject_sounds with inject_sounds

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+					playsound(loc,'sound/items/hypospray_long.ogg',50, 1, -1)
+				else
+					playsound(loc, pick(src.inject_sounds), 50, 1, -1)
+				log_attack("<font color='red'>[user.name] ([user.ckey]) sprayed [L.name] ([L.ckey]) with [src.name], which had [contained] (INTENT: [uppertext(user.a_intent)])</font>")
```
Replace src.name with just src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+	if(user.incapacitated() || !istype(user))
+		to_chat(user, "<span class='warning'>You can't do that right now!</span>")
+		return
+	if(src.contents == null)
```
No need for src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+		if("Cancel")
+			return
+		if("Inject")
+			src.mode = HYPO_INJECT
```
No need for src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+	set src in usr
+	if(usr.incapacitated())
+		return
+	var/choice = alert(usr, "Which application mode should this be? Current mode is: [src.mode ? "Spray" : "Inject"]", "", "Spray", "Cancel", "Inject")
```
No need for src

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,230 @@
+		if("Inject")
+			src.mode = HYPO_INJECT
+		if("Spray")
+			src.mode = HYPO_SPRAY
```
No need for src

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into hypospray-meme

Poojawa committed:
can I has real review now pls thx Bhijn

Poojawa committed:
pre-emptively avoiding a troll 'review'

jakeramsay007 reviewed (approved):
Code looks good, Bhijn's requested changes were made. Just waiting on Bhijn's review.

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,231 @@
+	to_chat(user, "[src] is set to [mode ? "Inject" : "Spray"] contents on application.")
+
+/obj/item/reagent_containers/hypospray/mkii/proc/unload_hypo(obj/item/reagent_containers/glass/bottle/vial/W, mob/user as mob)
+	if(contents == null)
```
if(!contents)

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,231 @@
+	if(user.incapacitated() || !istype(user))
+		to_chat(user, "<span class='warning'>You can't do that right now!</span>")
+		return
+	if(contents == null)
```
if(!contents)

deathride58 reviewed (dismissed):
Aside from a couple of small issues, I really don't notice any issues with the code.

You can probably move everything to modular_citadel for the sake of consistency, though

Poojawa committed:
WEH

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,238 @@
+	testing("forceMove valid, testing put_in_hands")
+	user.put_in_hands(vial)
+	testing("put_in_hands passed")
+	contents -= vial
```
Try using transferitemtoloc to move vials in and out of hyposprays instead of this

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,238 @@
+		to_chat(usr, "This Hypo needs to be loaded first!")
+		return
+	testing("Has hypo")
+	var/obj/item/reagent_containers/glass/bottle/vial/W
```
There's one issue

You're defining the var here, but you aren't setting it to anything

deathride58 commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,238 @@
+		user.visible_message("<span class='notice'>[user] begins loading a vial into \the [src].</span>","<span class='notice'>You start loading [vial] into \the [src].</span>")
+		if(!do_after(user,30) || vial || !(vial in user))
+			return FALSE
+		contents += vial
```
Pretty sure contents is already set by transferitemtoloc

Poojawa committed:
trying to code new functions is dumb

Poojawa committed:
reeeeee

put_in_hand doesn't exist apparently

and the shitty thing wants me to select mobs, not modes?

Poojawa committed:
ya lemmie just, rewrite everything

Poojawa committed:
fixes adaptation issues and sprites. Loading/unloading works

Now just to figure out reagent transfer_to code. aaaaaaaaaAAAAAAAAAAAAAAAAAAAAAAA

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into hypospray-meme

Poojawa committed:
HAHAHA YES

Poojawa committed:
Fully working with sounds and all.

Arokha commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,215 @@
+	if((istype(I, /obj/item/reagent_containers/glass/bottle/vial) && vial != null))
+		to_chat(user, "<span class='warning'>[src] can not hold more than one vial!</span>")
+		return FALSE
+	testing("we don't already have a vial in the hypo")
```
Shouldn't you remove this

Poojawa replied:
pfff. Stray testing verbs. pffff.

Probably.

Arokha commented on code/citadel/hypospraymkii.dm:
```diff
@@ -0,0 +1,215 @@
+			return
+		vial = V
+		reagents.maximum_volume = V.volume
+		V.reagents.trans_to(src, V.reagents.total_volume)
```
This not being duplicated in initialize is probably why vials like the CMO one need to be removed and readded, could put these few lines in their own eat_vial proc or something.

Poojawa replied:
Mayhaps, though I like the idea of adding a message saying basically 'the hypospray is still factory sealed!' even if it'd take more effort to make that flag a thing than to just fix it.

Poojawa committed:
remove a testy boi
